jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java

```diff
@@ -1175,6 +1114,20 @@ public void renewSessionId(String oldId, String oldExtendedId, String newId, Str
         {
             LOG.warn(e);
         }
+        finally
+        {
+            if (session != null)
+            {
+                try
+                {
+                    _sessionCache.release(newId, session);
+                }
+                catch (Exception e)
+                {
+                    LOG.warn(e);
+                }
+            }
+        }
```

Comment:
This is weird to me. Javadocs for `release()` say that the method should be called once a request is finished with a session.
It does not strike me that the request is finished with this session before existing this method. What am I missing?

Reply:
The javadoc on line 1103 says that the request reference count will be incremented by the call to sessionCache.renewSessionId;  the finally clause makes sure that this increment is decremented again. 

You have to consider the case where the session id exists on multiple different contexts, and thus must all be renewed when a request invokes changeId method.  There may be no active requests in any of the other contexts, thus sessioncache for each context needs to load in the session if its not already present - this has to increment a reference count. If we did not then decrement the count when we've finished changing the session id, the session would sit forever in the sessioncaches of these other contexts with ref count =1.  By always incrementing and decrementing for change of session id we maintain the correct behaviour of the caches.